File: lebanese_regulations/report/lebanese_general_ledger/lebanese_general_ledger.js
// Copyright (c) 2023, Your Name and contributors
// For license information, please see license.txt

frappe.query_reports["Lebanese General Ledger"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "finance_book",
            "label": __("Finance Book"),
            "fieldtype": "Link",
            "options": "Finance Book"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1,
            "width": "60px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1,
            "width": "60px"
        },
        {
            "fieldname": "account",
            "label": __("Account"),
            "fieldtype": "Link",
            "options": "Account",
            "get_query": function() {
                var company = frappe.query_report.get_filter_value('company');
                return {
                    "doctype": "Account",
                    "filters": {
                        "company": company,
                    }
                }
            }
        },
        {
            "fieldname": "voucher_no",
            "label": __("Voucher No"),
            "fieldtype": "Data",
            "width": "100px"
        },
        {
            "fieldname": "project",
            "label": __("Project"),
            "fieldtype": "Link",
            "options": "Project"
        },
        {
            "fieldname": "party_type",
            "label": __("Party Type"),
            "fieldtype": "Link",
            "options": "Party Type",
            "default": "",
            "on_change": function() {
                frappe.query_report.set_filter_value('party', "");
            }
        },
        {
            "fieldname": "party",
            "label": __("Party"),
            "fieldtype": "MultiSelectList",
            "get_data": function(txt) {
                if (!frappe.query_report.filters) return;

                let party_type = frappe.query_report.get_filter_value('party_type');
                if (!party_type) return;

                return frappe.db.get_link_options(party_type, txt);
            },
            "on_change": function() {
                var party_type = frappe.query_report.get_filter_value('party_type');
                var parties = frappe.query_report.get_filter_value('party');

                if(!party_type || parties.length === 0 || parties.length > 1) {
                    frappe.query_report.set_filter_value('party_name', "");
                    return;
                } else {
                    var party = parties[0];
                    var fieldname = erpnext.utils.get_party_name(party_type) || "name";
                    frappe.db.get_value(party_type, party, fieldname, function(value) {
                        frappe.query_report.set_filter_value('party_name', value[fieldname]);
                    });
                }
            }
        },
        {
            "fieldname": "party_name",
            "label": __("Party Name"),
            "fieldtype": "Data",
            "hidden": 1
        },
        {
            "fieldname": "group_by",
            "label": __("Group by"),
            "fieldtype": "Select",
            "options": ["", __("Group by Voucher"), __("Group by Account"), __("Group by Party")],
            "default": ""
        },
        {
            "fieldname": "cost_center",
            "label": __("Cost Center"),
            "fieldtype": "Link",
            "options": "Cost Center",
            get_query: () => {
                var company = frappe.query_report.get_filter_value('company');
                return {
                    filters: {
                        'company': company
                    }
                };
            }
        },
        {
            "fieldname": "show_in_lbp",
            "label": __("Show in LBP"),
            "fieldtype": "Check",
            "default": 1
        },
        {
            "fieldname": "show_foreign_currency",
            "label": __("Show Foreign Currency"),
            "fieldtype": "Check",
            "default": 1
        },
        {
            "fieldname": "include_dimensions",
            "label": __("Consider Accounting Dimensions"),
            "fieldtype": "Check",
            "default": 1
        },
        {
            "fieldname": "show_cancelled_entries",
            "label": __("Show Cancelled Entries"),
            "fieldtype": "Check"
        },
        {
            "fieldname": "include_default_book_entries",
            "label": __("Include Default Book Entries"),
            "fieldtype": "Check",
            "default": 1
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "debit" && data && data.debit > 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        else if (column.fieldname == "credit" && data && data.credit > 0) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        else if (column.fieldname == "debit_lbp" && data && data.debit_lbp > 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        else if (column.fieldname == "credit_lbp" && data && data.credit_lbp > 0) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        else if (column.fieldname == "debit_fc" && data && data.debit_fc > 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        else if (column.fieldname == "credit_fc" && data && data.credit_fc > 0) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        
        return value;
    },
    "tree": true,
    "name_field": "account",
    "parent_field": "parent_account",
    "initial_depth": 3
};

File: lebanese_regulations/report/lebanese_general_ledger/lebanese_general_ledger.py
# -*- coding: utf-8 -*-
# Copyright (c) 2023, Your Name and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, flt, cstr
from erpnext.accounts.report.general_ledger.general_ledger import execute as gl_execute

def execute(filters=None):
    """
    Execute the Lebanese General Ledger report
    
    Args:
        filters (dict): Report filters
        
    Returns:
        tuple: (columns, data)
    """
    # Add default filters for Lebanese GL
    if not filters:
        filters = {}
    
    # Add show_in_lbp filter if not present
    if "show_in_lbp" not in filters:
        filters["show_in_lbp"] = 1
    
    # Add show_foreign_currency filter if not present
    if "show_foreign_currency" not in filters:
        filters["show_foreign_currency"] = 1
    
    # Execute standard GL report
    columns, data = gl_execute(filters)
    
    # Modify columns and data for Lebanese requirements
    columns, data = modify_for_lebanese_requirements(filters, columns, data)
    
    return columns, data

def modify_for_lebanese_requirements(filters, columns, data):
    """
    Modify columns and data for Lebanese requirements
    
    Args:
        filters (dict): Report filters
        columns (list): Report columns
        data (list): Report data
        
    Returns:
        tuple: (columns, data)
    """
    # Add LBP and foreign currency columns if needed
    if filters.get("show_in_lbp"):
        # Add LBP columns
        columns = add_lbp_columns(columns)
        
        # Add LBP values to data
        data = add_lbp_values(data, filters)
    
    # Add foreign currency columns if needed
    if filters.get("show_foreign_currency"):
        # Add foreign currency columns
        columns = add_foreign_currency_columns(columns)
        
        # Add foreign currency values to data
        data = add_foreign_currency_values(data, filters)
    
    return columns, data

def add_lbp_columns(columns):
    """
    Add LBP columns to the report
    
    Args:
        columns (list): Report columns
        
    Returns:
        list: Modified columns
    """
    # Find the position to insert LBP columns
    debit_idx = -1
    credit_idx = -1
    
    for i, col in enumerate(columns):
        if col.get("fieldname") == "debit":
            debit_idx = i
        elif col.get("fieldname") == "credit":
            credit_idx = i
    
    if debit_idx >= 0 and credit_idx >= 0:
        # Insert LBP columns after debit and credit
        lbp_columns = [
            {
                "label": _("Debit (LBP)"),
                "fieldname": "debit_lbp",
                "fieldtype": "Currency",
                "options": "LBP",
                "width": 120
            },
            {
                "label": _("Credit (LBP)"),
                "fieldname": "credit_lbp",
                "fieldtype": "Currency",
                "options": "LBP",
                "width": 120
            }
        ]
        
        # Insert after credit column
        columns.insert(credit_idx + 1, lbp_columns[1])
        columns.insert(debit_idx + 1, lbp_columns[0])
    
    return columns

def add_foreign_currency_columns(columns):
    """
    Add foreign currency columns to the report
    
    Args:
        columns (list): Report columns
        
    Returns:
        list: Modified columns
    """
    # Find the position to insert foreign currency columns
    debit_idx = -1
    credit_idx = -1
    
    for i, col in enumerate(columns):
        if col.get("fieldname") == "debit":
            debit_idx = i
        elif col.get("fieldname") == "credit":
            credit_idx = i
    
    if debit_idx >= 0 and credit_idx >= 0:
        # Insert foreign currency columns
        fc_columns = [
            {
                "label": _("Foreign Currency"),
                "fieldname": "foreign_currency",
                "fieldtype": "Link",
                "options": "Currency",
                "width": 100
            },
            {
                "label": _("Exchange Rate"),
                "fieldname": "exchange_rate",
                "fieldtype": "Float",
                "precision": 6,
                "width": 100
            },
            {
                "label": _("Debit (FC)"),
                "fieldname": "debit_fc",
                "fieldtype": "Currency",
                "options": "foreign_currency",
                "width": 120
            },
            {
                "label": _("Credit (FC)"),
                "fieldname": "credit_fc",
                "fieldtype": "Currency",
                "options": "foreign_currency",
                "width": 120
            }
        ]
        
        # Insert before debit column
        for i, col in enumerate(fc_columns):
            columns.insert(debit_idx + i, col)
    
    return columns

def add_lbp_values(data, filters):
    """
    Add LBP values to the report data
    
    Args:
        data (list): Report data
        filters (dict): Report filters
        
    Returns:
        list: Modified data
    """
    for row in data:
        if not row.get("posting_date"):
            continue
        
        # Get exchange rate
        exchange_rate = 1.0
        if row.get("account_currency") and row.get("account_currency") != "LBP":
            exchange_rate = get_exchange_rate(row.get("account_currency"), "LBP", row.get("posting_date"))
        
        # Calculate LBP values
        row["debit_lbp"] = flt(row.get("debit")) * flt(exchange_rate)
        row["credit_lbp"] = flt(row.get("credit")) * flt(exchange_rate)
    
    return data

def add_foreign_currency_values(data, filters):
    """
    Add foreign currency values to the report data
    
    Args:
        data (list): Report data
        filters (dict): Report filters
        
    Returns:
        list: Modified data
    """
    for row in data:
        if not row.get("posting_date"):
            continue
        
        # Set foreign currency values
        if row.get("account_currency") and row.get("account_currency") != "LBP":
            row["foreign_currency"] = row.get("account_currency")
            row["exchange_rate"] = get_exchange_rate(row.get("account_currency"), "LBP", row.get("posting_date"))
            
            # Calculate foreign currency values
            if row.get("debit") > 0:
                row["debit_fc"] = flt(row.get("debit_in_account_currency"))
                row["credit_fc"] = 0
            else:
                row["debit_fc"] = 0
                row["credit_fc"] = flt(row.get("credit_in_account_currency"))
        else:
            # For LBP accounts, set foreign currency values to zero
            row["foreign_currency"] = ""
            row["exchange_rate"] = 1.0
            row["debit_fc"] = 0
            row["credit_fc"] = 0
    
    return data

def get_exchange_rate(from_currency, to_currency, date):
    """
    Get exchange rate between two currencies
    
    Args:
        from_currency (str): From currency
        to_currency (str): To currency
        date (str): Date for exchange rate
        
    Returns:
        float: Exchange rate
    """
    if from_currency == to_currency:
        return 1.0
    
    # Get exchange rate from Currency Exchange
    exchange_rate = frappe.db.get_value("Currency Exchange",
        {
            "from_currency": from_currency,
            "to_currency": to_currency,
            "date": ("<=", date)
        },
        "exchange_rate",
        order_by="date desc"
    )
    
    if not exchange_rate:
        # Try reverse lookup
        exchange_rate = frappe.db.get_value("Currency Exchange",
            {
                "from_currency": to_currency,
                "to_currency": from_currency,
                "date": ("<=", date)
            },
            "exchange_rate",
            order_by="date desc"
        )
        
        if exchange_rate:
            exchange_rate = 1.0 / flt(exchange_rate)
    
    return exchange_rate or 1.0